File: frontend/src/api/model4_alignment.ts
import apiClient from './apiClient';

export interface GoalAlignmentRequest {
  goal_text: string;
  html_mastery: number;
  css_mastery: number;
  js_mastery: number;
  react_mastery: number;
  python_mastery: number;
  ml_mastery: number;
  dsa_mastery: number;
  environment: string;
  engagement_score: number;
  consistency_score: number;
  integrity_score: number;
  anomaly_score: number;
}

export interface GoalAlignmentResponse {
  student_id: number;
  goal_text: string;
  goal_type: string;
  goal_specificity_score: number;
  collaboration_score: number;
  alignment_score: number;
  predicted_domain: string;
  all_domain_scores: Record<string, number>;
  skill_gap: number;
  skill_gap_vector: Record<string, number>;
  weakest_topics: string[];
  context_adjustment_score: number;
  learning_mode_hint: string;
  integrity_flag: boolean;
  scaffold_level: string;
  behavior_summary: string;
  recommendation: string;
  learning_path: string[];
  resources: string[];
  explanation: string;
  confidence_score: number;
}

export interface GoalProfileOut {
  id: number;
  student_id: number;
  goal_text: string;
  goal_type: string | null;
  goal_specificity_score: number | null;
  alignment_score: number | null;
  predicted_domain: string | null;
  skill_gap: number | null;
  weakest_topics: string[] | null;
  scaffold_level: string | null;
  learning_mode_hint: string | null;
  integrity_flag: boolean | null;
  recommendation: string | null;
  learning_path: string[] | null;
  resources: string[] | null;
  confidence_score: number | null;
  behavior_summary: string | null;
  explanation: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface MasteryPrefillOut {
  html_mastery: number;
  css_mastery: number;
  js_mastery: number;
  react_mastery: number;
  python_mastery: number;
  ml_mastery: number;
  dsa_mastery: number;
  source_topics: Record<string, string[]>;
  has_data: boolean;
}

export const alignmentApi = {
  analyzeGoal: async (studentId: number, data: GoalAlignmentRequest): Promise<GoalAlignmentResponse> => {
    const response = await apiClient.post(`/alignment/analyze/${studentId}`, data);
    return response.data;
  },

  getProfile: async (studentId: number): Promise<GoalProfileOut> => {
    const response = await apiClient.get(`/alignment/profile/${studentId}`);
    return response.data;
  },

  getMasteryPrefill: async (studentId: number): Promise<MasteryPrefillOut> => {
    const response = await apiClient.get(`/alignment/mastery-prefill/${studentId}`);
    return response.data;
  },

  analyzeBatch: async (studentIds: number[]): Promise<GoalAlignmentResponse[]> => {
    const response = await apiClient.post('/alignment/analyze-batch', { student_ids: studentIds });
    return response.data;
  },

  getAllProfiles: async (): Promise<GoalProfileOut[]> => {
    const response = await apiClient.get('/alignment/all');
    return response.data;
  },
};

File: backend/app/schemas/model4_alignment.py
"""
model4_alignment.py
Pydantic Schemas — Model 4 (Purpose & Skill Alignment)
"""

import json
from pydantic import BaseModel, Field, model_validator
from typing import Optional, List, Dict, Any
from datetime import datetime


class GoalAlignmentRequest(BaseModel):
    """Request body for POST /alignment/analyze"""
    goal_text: str = Field(..., min_length=3, description="Student's goal statement")

    # Skill mastery (0.0 – 1.0)
    html_mastery:   float = Field(0.0, ge=0.0, le=1.0)
    css_mastery:    float = Field(0.0, ge=0.0, le=1.0)
    js_mastery:     float = Field(0.0, ge=0.0, le=1.0)
    react_mastery:  float = Field(0.0, ge=0.0, le=1.0)
    python_mastery: float = Field(0.0, ge=0.0, le=1.0)
    ml_mastery:     float = Field(0.0, ge=0.0, le=1.0)
    dsa_mastery:    float = Field(0.0, ge=0.0, le=1.0)

    # Learning context
    environment:       str   = Field("online", description="online/lab/project/self-study")
    engagement_score:  float = Field(0.5, ge=0.0, le=1.0)
    consistency_score: float = Field(0.5, ge=0.0, le=1.0)
    integrity_score:   float = Field(0.8, ge=0.0, le=1.0)
    anomaly_score:     float = Field(0.0, ge=0.0, le=1.0)


class GoalAlignmentResponse(BaseModel):
    """Full response from the Model 4 pipeline"""
    student_id:               int
    goal_text:                str
    goal_type:                str
    goal_specificity_score:   float
    collaboration_score:      float

    alignment_score:          float
    predicted_domain:         str
    all_domain_scores:        Dict[str, float]

    skill_gap:                float
    skill_gap_vector:         Dict[str, float]
    weakest_topics:           List[str]

    context_adjustment_score: float
    learning_mode_hint:       str
    integrity_flag:           bool
    scaffold_level:           str
    behavior_summary:         str

    recommendation:           str
    learning_path:            List[str]
    resources:                List[str]
    explanation:              str
    confidence_score:         float


class GoalProfileOut(BaseModel):
    """
    Stored goal profile returned by GET /alignment/profile/{student_id}.
    JSON string columns (weakest_topics, learning_path, resources) are
    automatically parsed into proper Python lists on read.
    """
    id:                       int
    student_id:               int
    goal_text:                str
    goal_type:                Optional[str]
    goal_specificity_score:   Optional[float]
    alignment_score:          Optional[float]
    predicted_domain:         Optional[str]
    skill_gap:                Optional[float]
    weakest_topics:           Optional[List[str]]
    scaffold_level:           Optional[str]
    learning_mode_hint:       Optional[str]
    integrity_flag:           Optional[bool]
    recommendation:           Optional[str]
    learning_path:            Optional[List[str]]
    resources:                Optional[List[str]]
    confidence_score:         Optional[float]
    behavior_summary:         Optional[str]
    explanation:              Optional[str]
    created_at:               Optional[datetime]
    updated_at:               Optional[datetime]

    @model_validator(mode="before")
    @classmethod
    def parse_json_fields(cls, values):
        """
        SQLAlchemy ORM objects expose attributes, not dict keys.
        Convert the ORM object to a dict first, then parse JSON string columns.
        """
        # Handle both ORM objects and plain dicts
        if not isinstance(values, dict):
            values = {c.key: getattr(values, c.key) for c in values.__table__.columns}

        for field in ("weakest_topics", "learning_path", "resources"):
            raw = values.get(field)
            if isinstance(raw, str):
                try:
                    values[field] = json.loads(raw)
                except (json.JSONDecodeError, ValueError):
                    values[field] = []
        return values

    class Config:
        from_attributes = True


class MasteryPrefillOut(BaseModel):
    """
    Pre-filled skill mastery values derived from knowledge tracing states.
    Returned by GET /alignment/mastery-prefill/{student_id}.
    source_topics shows which topics contributed to each skill score.
    """
    html_mastery:    float
    css_mastery:     float
    js_mastery:      float
    react_mastery:   float
    python_mastery:  float
    ml_mastery:      float
    dsa_mastery:     float
    source_topics:   Dict[str, List[str]]   # skill → list of matched topic titles
    has_data:        bool                   # False if no knowledge states exist yet


class BatchAnalyzeRequest(BaseModel):
    """Request body for POST /alignment/analyze-batch"""
    student_ids: List[int] = Field(..., min_length=1, description="List of student IDs to re-analyze")
